readme.txt
=== SG What's Playing ===
Contributors: vmole
Tags: music, sidebar
Requires at least: 2.2.0
Tested up to: 2.3
Stable tag: trunk

Display currently playing song from MPD or other music players in a sidebar widget.

== Description ==

Yes, it's yet another "currently playing" plugin. What makes this one
better (or, at least, different)?

* Support for [MPD](http://www.musicpd.org/ "Music Player Daemon")

* Support for others is possible and possibly easy - see below.

* No file transfer -- WordPress is updated via HTTP POST.

* Updater should work on Linux/BSD, OSX, and Windows.

SG What's Playing (SGWP from now on) is comes in two components: the
WordPress plugin, and a [Python](http://www.python.org) program
(`sgwpupdate.py`) that reads data from the player and POSTs it to the
blog.

The `sgwpupdate.py` program can be run as a one-shot or, on Unix-like
systems, as a daemon. In one-shot mode, you can specify song details
rather than reading them from the player. Thus, if your player can
trigger a program on song or status change (e.g. Audacious, WinAMP (I
think!)), you can probably use `sgwpupdate.py` to update your
WordPress blog.

In daemon mode, `sgwpupdate.py` uses a player-specific module to
obtain the song meta-data from the player and then sends the data to
your WordPress blog. At present, the only included module is for MPD,
but writing modules for other players shouldn't be too difficult. See
the "Custom Player Modules" section below.

== Installation ==

= The WordPress plugin and widget. =

* Unpack the archive.

* Copy `sgwhatsplaying.php` and `widget.php` to a `sgwhatsplaying`
  directory in your WordPress plugins directory. Or just unpack the
  archive under your plugins directory - all the python stuff will be
  ignored.

* Activate the plugin.

* Goto Options->SGWhatsPlaying, and set the "password". Don't use a
  password you use for any real access - this is just key to keep
  random losers from screwing with your blog, and is not secure in any
  real way.

= The updater program =

* Unpack the archive on the machine with access to your music player.

* If you don't have Python installed, install it. 

* (OPTIONAL) Run `python setup.py install` to install the program and
  module into your regular Python site directories. But you can skip
  this: `sgwpupdate.py` will run fine out of any directory.

* Copy sgwpupdate.conf to your `$HOME/.sgwpupdate.conf`, or
  `/etc/sgwpupdate.conf`.

* Edit the configuration file to set `password` and `blog` in the
  `sgwpupdate` section, and (if using MPD) `mpdhost` and `mpdport` in
  the `sgwpmpd` section. See later in this document for complete
  configuration documentation.

* Run `sgwpupdate.py --debug` to test. If you're not using MPD, you can
  test with `sgwpupdate.py --debug --title 'Some Song Title' --artist 
  'Some Artist' --album 'Album Name' --state 'play'`.

* To start in daemon mode, run `sgwpupdate.py --daemon`.

* The file `sgwupdate.ini` is a Debian-compatible initrc file. Copy it
  to `/etc/init.d/sgwpupdate` and create the desired links in
  `/etc/rc?.d` to have sgwpupdate.py automatically start on system
  boot.

== Screenshots ==

Not yet.

== Configuration ==

Configuration is via file. On Unix-like systems (includding OSX?), the
file is read from `~/.sgwpupdate.conf` or `/etc/sgwpupdate.conf`. If
both exist, entries in the former override those in the latter. The
file format is the usual WIN.INI style, with sections denoted by
square brackets and items specified as `item = value`. Normally, the
file will have two sections: the standard `[sgwpupdate]` section and a
module specific section such as `[sgwpmpd]`.

The supported configuration items in `[sgwpupdate]` section are:

* `blog` - URL to WordPress blog, such as `http://www.example.net/wordpress`

* `password` - simple password to allow updates, must match option in
  plugin configuration.

* `module` - name of python module to do player specific update.

* `logfile` - full path to logfile (optional)

* `pidfile` - full path to name of file where PID will be written, if
  running in daemon mode. (optional)

The supported configuration items for the `[sgwpmpd]` module section are:

* `mpdhost` - name of host where MPD is running, eg `localhost` or
  `mpd.example.net`

* `mpdport` - TCP port number for MPD. Almost always 6600.

== Custom Player Modules ==

Write a Python module containing a function name `getsong` that takes
a single argument, which will be a dictionary of the config items from
the module-specific configuration section. The function should return
a tuple of (remainingtime, songdata). The `remainingtime` value sets
how long `sgwpupdate.py` will wait before checking again in daemon
mode. The dictionary `songdata` should contain the following keys:

* `artist` - song performer or writer.

* `title` - song title.

* `album` - album title for the song.

* `state` - current state of the player, should be one of 'play',
  'pause', or 'stop'.

Your best bet is probably to copy `sgwpmpd.py` and work from there.


Changelog
2007-10-12  Release 1.1

	* sgwhatsplaying.php:
	Update version to 1.1
	[bc957bbefa77]

	* sgwpupdate.py:
	Don't crash when remote server isn't available.
	[b07afd9c61c1]

	* widget.php:
	Don't show 'no recent updates' in pause mode, because sgwpupdate
	doesn't send new info without changes.
	[908476e6da86]

	* sgwhatsplaying.php:
	Fix for WP 2.3
	[cbf0db8fb0e5]



sgwpmpd.py
#!/usr/bin/python

"""SGWhatsPlaying module for MPD

"""

import sys
import logging
import socket
import telnetlib

def mpd_connect(host="localhost", port=6600):
    """Connect to MPD server
    """
    logging.debug("Connecting to MPD at %s:%d", host, port)
    try:
        tn = telnetlib.Telnet(host, port)
    except socket.error, inst:
        logging.error("connection to %s:%d failed: %s", host, port, inst)
        return None
    
    (mid, mo, f) = tn.expect(["^OK MPD .*"], 5)
    logging.debug("tnconnect returned: %s ", f)
    return tn


def rspdict(f):
    """Convert multiple lines returned from MPD into dictionary

    Assumed form is "key: value", other lines are ignored
    """
    resp=f.splitlines()
    result = {}
    for l in resp[:-1]:
        logging.debug("rspdict: %s", l)
        try:
            (k, v) = l.split(": ")
            result[k] = v
        except ValueError:
            pass # ignore bad splits.
    return result
    

def mpdstatus(mpd):
    mpd.write("status\r\n")
    (mid, mo, f) = mpd.expect(["OK"], 5)
    resp=f.splitlines()
    mpds = rspdict(f)
    return mpds

def mpdcurrent(mpd):
    mpd.write("currentsong\r\n")
    (mid, mo, f) = mpd.expect(["OK"], 5)
    mpdc = rspdict(f)
    return mpdc

def getsong(config):
    """Return (time_to_next, wpdata) tuple

    Parameters
    
    - `config` dict of values from modules config file section
    """
    mpd = mpd_connect(config['mpdhost'], int(config['mpdport']))
    if not mpd:
        logging.error("MPD unavailable")
        return None
    
    mpds = mpdstatus(mpd)
    mpdc = mpdcurrent(mpd)
    mpd.close()

    wpdata = {}
    wpdata['artist'] = mpdc.get('Artist', 'unknown')
    wpdata['title'] = mpdc.get('Title', 'unknown')
    wpdata['album'] = mpdc.get('Album', 'unknown')
    wpdata['state'] = mpds.get('state', 'unknown')
    if 'time' in mpds:
        (p, t) = mpds['time'].split(":")
        p = int(p)
        t = int(t)
        rt =  t - p + 1
    else:
        rt = 60
    return (rt, wpdata)




widget.php
<?php
/*  Copyright 2007  Steve Greenland  (email : steveg@moregruel.net)

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/


/* The SG What's Playing widget for the sidebar */

function widget_SGWhatsPlaying_init() {

	if ( !function_exists('register_sidebar_widget') )
		return;

function widget_SGWhatsPlaying($args) {
	global $wpdb;

	// "$args is an array of strings that help widgets to conform to
	// the active theme: before_widget, before_title, after_widget,
	// and after_title are the array keys." - These are set up by the theme
	extract($args);

	// Get the current data
	$table_name = $wpdb->prefix . "sgwhatsplaying";
	$sql = "SELECT * FROM " . $table_name . " WHERE id=1 LIMIT 1";
	$songs = $wpdb->get_results($sql);
	$song = $songs[0];

	$timeago = time() - strtotime($song->time); /* Time since last update, in seconds */

	echo $before_widget . $before_title . "Now Playing";
	echo $after_title;
	echo "<ul><li>";
	if ($song->state == 'stop') {
		echo "Nothing's Playing...";
	} else {
		echo stripslashes($song->artist);
		echo " - ";
		echo stripslashes($song->title);
		if ($song->state == 'pause') {
			echo " (paused)";
		} elseif ($timeago > 30 * 60) {
			echo " (no recent updates)";
		}
	}
	echo "</li></ul>";

	echo $after_widget;
}

function widget_SGWhatsPlaying_control() {
	// No options.
	echo "<div>No widget options</div>";
}

	register_sidebar_widget(array("SG What's Playing", 'widgets'), 'widget_SGWhatsPlaying');
	register_widget_control(array("SG What's Playing", 'widgets'), 'widget_SGWhatsPlaying_control', 300, 200);

}
?>


sgpyutil/daemon.py
#!/usr/bin/python

"""Unix Daemon creation and management

Functions to daemonize a python program, and tools to find and kill
that daemon later.

This is an abbreviated and modified version of a python daemon recipie
at http://aspn.activestate.com/ASPN/Cookbook/Python/Recipe/278731
posted by Chad Schroeder. Thanks, Chad.

His code was nicely commented, I stripped it out for compactness sake.
    
"""

import sys
import os

def maxfd(MAXFD=1024):
    """Attempt to retrieve maximum system fd.

    This is one of those nasty values that is system dependent, may be
    dynamic, and may not exist at all.

    """
    import resource		# Resource usage information.
    maxfd = resource.getrlimit(resource.RLIMIT_NOFILE)[1]
    if (maxfd == resource.RLIM_INFINITY):
        maxfd = MAXFD
  
    return maxfd
    

def daemonize(closefds=(0,1,2), pidfile=None, workdir="/", umask=0):
    """Daemonize the current process.

    - closefds - list of file descriptors to be closed.

    - pidfile - name of pidfile to be created.

    - workdir - working directory for daemon, usually "/".

    - umask - default umask for new files, usually 0
    
    """


    try:
        pid = os.fork()
    except OSError, e:
        raise Exception, "%s [%d]" % (e.strerror, e.errno)

    if (pid != 0):
        os._exit(0)	# Exit parent of the first child.

    os.setsid()
    try:
        pid = os.fork()	# Fork a second child.
    except OSError, e:
        raise Exception, "%s [%d]" % (e.strerror, e.errno)

    if (pid != 0):
        # We might want to write the pid file here.
        if pidfile:
            os.umask(umask)
            f = open(pidfile, "w")
            f.write("%d\n" % (pid))
            f.close()
        os._exit(0)	# Exit parent of the second child.
    
    # This is our daemon process.
    if workdir:
        os.chdir(workdir)
    # Iterate through and close all file descriptors.
    for fd in closefds:
        try:
            os.close(fd)
        except OSError:	# ERROR, fd wasn't open to begin with (ignored)
            pass
            
    # Just close the stdio fds
    if (hasattr(os, "devnull")):
        REDIRECT_TO = os.devnull
    else:
        REDIRECT_TO = "/dev/null"
        
    os.open(REDIRECT_TO, os.O_RDWR)	# standard input (0)
    os.dup2(0, 1)			# standard output (1)
    os.dup2(0, 2)			# standard error (2)
    
    return(0)


from optparse import OptionParser

def main(argv=None):
    # Option parsing
    parser = OptionParser(usage="usage: daemon.py [options]", version="0.1")
    parser.add_option("-d", "--debug", action="store_true", dest="debug",
                      default=False,
                      help="print program activity")
    parser.add_option("-w", "--working-directory", dest="workdir", default=None,
                      help="Daemon working directory")
    parser.add_option("-u", "--umask", dest="umask", default=0022, type="int",
                      help="new umask")

    parser.add_option("-p", "--pidfile", dest="pidfile", default=None,
                      help="file in which to store PID")
    
    if argv is None:
        argv = sys.argv

    (options, args) = parser.parse_args()

    daemonize(workdir=options.workdir,
              umask=options.umask,
              pidfile=options.pidfile)

    return 0

if __name__ == "__main__":
    sys.exit(main())


sgwhatsplaying.php
<?php
/*
Plugin Name: SG What's Playing
Plugin URI: http://www.moregruel.net/?page_id=73
Description: Display currently playing song in sidebar or posts
Author: Steve Greenland <steveg@moregruel.net>
Version: 1.1
Author URI: http://www.moregruel.net/
*/

/*  Copyright 2007  Steve Greenland  (email : steveg@moregruel.net)

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

/* Plugin constants */

define('SGWP_DB_VERSION', '1');


/* Create DB and options on install */

function sgwhatsplaying_install() {
	global $wpdb;

	$table_name = $wpdb->prefix . "sgwhatsplaying";

	if ($wpdb->get_var("SHOW TABLES LIKE '$table_name'") != $table_name) {
		$sql = "CREATE TABLE " . $table_name . " (
		id TINYINT,
	  	time TIMESTAMP,
	  	artist CHAR(30),
	  	title CHAR(30),
	  	album CHAR(30),
	  	state CHAR(10)
		);";

		if( file_exists( ABSPATH . 'wp-admin/includes/upgrade.php' ))
			require_once(ABSPATH . 'wp-admin/includes/upgrade.php');
		else
			require_once(ABSPATH . 'wp-admin/upgrade-functions.php');

		dbDelta($sql);

		$coreid = 1;
		$insert = "INSERT INTO " . $table_name . "(id) " .
       		     "VALUES (" . $coreid. ")";

		$results = $wpdb->query( $insert );

      		add_option("sgwhatsplaying_db_version", SGWP_DB_VERSION);
      		add_option("sgwhatsplaying_password", "", "Update password");
	}

}

/* Update the DB with new info */

function sgwhatsplaying_update($myarg) {
	global $wpdb;
	if ($_POST['sgwhatsplaying'] == 'update') {
		$pwd = $_POST['password'];
		$rpwd = get_option("sgwhatsplaying_password");
		if ($pwd != $rpwd) {
			echo "FAIL";
			echo " bad password";
			die(0);
		}
		$table_name = $wpdb->prefix . "sgwhatsplaying";
		$artist = $_POST['artist'];
		$album = $_POST['album'];
		$title = $_POST['title'];
		$playstate = $_POST['state'];
		$sql = "UPDATE $table_name " .
			"SET artist = '" . $wpdb->escape($artist) . "', " .
			"album = '" . $wpdb->escape($album) . "', " .
			"title = '" . $wpdb->escape($title) . "', " .
			"state = '" . $wpdb->escape($playstate) . "' " .
			"WHERE id = 1";

		$results = $wpdb->query( $sql );
		if ($results == 1) {
			echo "OK";
		} else {
			echo "FAIL";
			$wpdb->print_error();
		}
		die(0);
	}
}

/* Option page */

function sgwhatsplaying_options() {
	if ('POST' == $_SERVER['REQUEST_METHOD']) {
		if (isset($_POST['sgwp_password'])) {
			update_option('sgwhatsplaying_password', $_POST['sgwp_password']);
		}
	}

	$sgwp_password = get_option('sgwhatsplaying_password');

	echo <<<END1
<div class="wrap">
	<h2>SG What's Playing Options</h2>

	<form action="" method="post" id="sgwhatsplaying-conf">
	<fieldset class="options">
	<ul><li>Password for sgwpupdate: <input type=text name="sgwp_password" value="$sgwp_password" size=16/>
	</li></ul></fieldset>
	<p class="submit">
		<input type="submit" name="Submit" value="Update Options " />
	</p>
	</form>
	</div>
END1;
}

function sgwhatsplaying_options_activate() {
	add_options_page("SG What's Playing Options", "SGWhatsPlaying", 8,
			"testoptions", "sgwhatsplaying_options");
}

/* The widget for the sidebar */

require_once('widget.php');

/* Okay, hook everything up into WP */
add_action('activate_sgwhatsplaying/sgwhatsplaying.php',
	 'sgwhatsplaying_install');
add_action('admin_menu', 'sgwhatsplaying_options_activate');
add_action('widgets_init', 'widget_SGWhatsPlaying_init');
add_action('init', 'sgwhatsplaying_update');

?>


sgwpupdate.py
#!/usr/bin/python

"""Update SG What's Playing Plugin Wordpress blog

Config file has following sections

[sgwpupdate]
blog=URL
password=string
module=modname

[modname]
module_specific opttions.

for example:

[sgwpmpd]
mpdhost=hostname
mpdport=portnum

"""

import sys
import os
import time
from optparse import OptionParser
import ConfigParser

import logging, logging.handlers
import socket
import telnetlib

import urllib

from sgpyutil import daemon

def update_sgwhatsplaying(wpblogurl, wppassword,  wpdata):
    """wpdata is a dict with keys "artist", "title", "album", and "state"

    'state' should one of ("play", "pause", "stop")
    """
    logging.info("Updating %s with '%s', '%s', '%s', and '%s'",
                 wpblogurl,
                 wpdata['state'],
                 wpdata['artist'],
                 wpdata['title'],
                 wpdata['album'])
    params = wpdata.copy()
    params.update({'sgwhatsplaying':'update'})
    params.update({'password':wppassword})
    logging.debug("Parameters set as %s", params)
    params = urllib.urlencode(params)
    logging.debug("Parameters encoded as %s", params)
    try:
        f = urllib.urlopen(wpblogurl, params)
        logging.debug("url returned:\n %s", f.info())
        logging.debug(f.read().splitlines()[:30])
    except:
        logging.info("Server at %s unavailable", wpblogurl)
    return


def main(argv=None):
    # Option parsing
    parser = OptionParser(usage="usage: %prog [options] host version", version="0.1")
    parser.add_option("-d", "--debug", help="print program activity",
                      action="store_true", dest="debug", default=False)
    parser.add_option("-t", "--test", help="Show results of song lookup, but don't update blog",
                      action="store_true", dest="test", default=False)
    
    parser.add_option("-D", "--daemonize", help="Run as Unix daemon",
                      action="store_true", dest="daemon", default=False)

    parser.add_option("", "--title", help="Song title", action="store",
                      default="")
    parser.add_option("", "--album", help="Song title", action="store",
                      default="")
    parser.add_option("", "--artist", help="Song title", action="store",
                      default="")
    parser.add_option("", "--state", help="Song title", action="store",
                      default="")

    if argv is None:
        argv = sys.argv

    # Read command line options
    (options, args) = parser.parse_args()

    # Set up the console logger. We'll have two - one to the console,
    # one to a file, but we can't setup the file one until we've read
    # the config file
    
    logging.getLogger().setLevel(logging.NOTSET) # Handlers control log level
    conlog = logging.StreamHandler()
    conlog.setFormatter(logging.Formatter('%(levelname)-8s %(message)s'))
    if (options.debug):
        conlog.setLevel(logging.DEBUG)
    else:
        conlog.setLevel(logging.WARNING)
    logging.getLogger().addHandler(conlog)

    # Read config file
    cf = ConfigParser.ConfigParser()
    cfiles = cf.read(["/etc/sgwpupdate.conf", os.path.expanduser("~/.sgwpupdate.conf")])
    logging.debug("Read config files: %s", cfiles)

    # Convert config file to dict of dict.
    config = {}
    for s in cf.sections():
        config[s] = dict(cf.items(s))

    # Convenience configuration variables - one for the main, one for the
    # module
    cfmain = config['sgwpupdate']
    if 'module' in cfmain:
        cfmod = config[cfmain['module']]
    
    # make sure we have required info
    rk = ['blog', 'password'];
    for k in cfmain.keys():
        if k in rk:
            rk.remove(k)
    if rk:
        logging.error("Missing configuration values for:%s", " ".join(rk))
        return 2

    if 'logfile' in cfmain:
        filelog = logging.handlers.TimedRotatingFileHandler(
            filename = cfmain['logfile'],
            when="midnight", backupCount=2)
        filelog.setFormatter(logging.Formatter('%(asctime)s %(levelname)-8s %(message)s'))
        if options.debug:
            filelog.setLevel(logging.DEBUG)
        else:
            filelog.setLevel(logging.INFO)
            
        logging.getLogger().addHandler(filelog)
    
    songmodname =  cfmain['module']
    logging.debug("Loading song module %s", songmodname)
    songmod = __import__(songmodname)


    # Validate arguments.
    if options.daemon and options.title:
        logging.error("Option --daemon conflicts with --title")
        return 2


    # Okay, set for main loop 

    if options.daemon:
        daemon.daemonize(pidfile=cfmain.get('pidfile', None))
        logging.getLogger().removeHandler(conlog)

    wpdata_prev = {}
    wpdata = {}
    while 1:
        if options.title:
            wpdata['title'] = options.title
            wpdata['album'] = options.album
            wpdata['artist'] = options.artist
            wpdata['state'] = options.state
            dtime = -1
        else:
            (dtime, wpdata) = songmod.getsong(cfmod)
    
        if wpdata and wpdata != wpdata_prev: 
            if options.test:
                print "Artist: ", wpdata['artist']
                print "Title: ", wpdata['title']
                print "Album: ", wpdata['album']
                print "Player state: ", wpdata['state']
                logging.info("test: %(artist)s - %(title)s from %(album)s,"
                             " player is in %(state)s" % wpdata)
                return 0
            wpdata_prev = wpdata.copy()
            wpdata.update(dtime=dtime)
            logging.debug("wpdata: artist='%(artist)s', title='%(title)s',"
                          " album='%(album)s', "
                          " state='%(state)s', dtime=%(dtime)d" % wpdata)
                
            update_sgwhatsplaying(cfmain['blog'],
                                  cfmain['password'],
                                  wpdata)
        else:
            logging.debug("songmod.getsong() returned None for wpdata")
            
        if not options.daemon:
            return 0
        
        if (dtime > 0):
            time.sleep(dtime)
        else:
            time.sleep(60)

    return 0



if __name__ == "__main__":
    sys.exit(main())
